topological_action_planner/scripts/tap.py

```diff
@@ -0,0 +1,149 @@
+from geometry_msgs.msg import PoseStamped, Pose, Point, Quaternion
+from std_msgs.msg import Header
+import tf2_ros
+import pykdl_ros
+
+from ed_py.utility import rooms_of_volume, rooms_of_entity
+from ed_py.world_model import WM
+from topological_action_planner.serialisation import from_dicts
+from topological_action_planner.planning import TopoPlanner
+from topological_action_planner.util import visualize, generate_dummy_graph
+from topological_action_planner.visualisation import create_tap_marker_array
+from cb_base_navigation_msgs.srv import GetPlan, GetPlanRequest, GetPlanResponse
+from ed_navigation_msgs.srv import GetGoalConstraint
+from cb_base_navigation_msgs.msg import PositionConstraint
+
+
+class TopoPlannerNode:
+    def __init__(self):
+        self.G = from_dicts(rospy.get_param("~edges"))
+        self.plot = rospy.get_param("~plot", False)
+
+        # TODO: These are magic numbers. Can be parametrized, but these are the defaults
+        action_costs = rospy.get_param(
+            "~action_costs",
+            {
+                Edge.ACTION_DRIVE: 1,  # per meter distance driven
+                Edge.ACTION_OPEN_DOOR: 5,  # per door opened
+                Edge.ACTION_PUSH_OBJECT: 10,  # per item pushed
+            },
+        )
+
+        self._wm = WM()
+        self._topo_planner = TopoPlanner(self._wm, action_costs)
+
+        self._get_constraint_srv = rospy.ServiceProxy("ed/navigation/get_constraint", GetGoalConstraint)
+
+        self._srv_plan = rospy.Service("~get_plan", Plan, self._srv_plan_cb)
+        self._srv_update_edge = rospy.Service("~update_edge", UpdateEdge, self._srv_update_edge_cb)
+        self._pub_grasp_marker = rospy.Publisher("~graph", MarkerArray, queue_size=10, latch=True)
+
+        # self.G = generate_dummy_graph()
+        # to_yaml(self.G, '/tmp/graph.yaml')
+        if self.plot:
+            visualize(self.G)
+        self._pub_grasp_marker.publish(create_tap_marker_array(self.G, self._wm))
+
+        rospy.loginfo("Topological action planner started")
+
+    def _srv_plan_cb(self, req: PlanRequest) -> PlanResponse:
+        """
+        Query the graph for a plan from an origin to some destination.
+        If the origin is not given, interpret that as starting from the robot's current position.
+        """
+        graph = copy.deepcopy(self.G)  # type: nx.Graph
+
+        for node in graph.nodes.keys():
+            node0_entity = self._wm.get_entity(node[0])
+            graph.nodes[node]["room"] = rooms_of_volume(
+                self._wm, node0_entity, node[1]
+            )  # Based on where they are in ED?
+
+        if req.origin.entity == "":
+            # This indicates the plan starts from the robot's current pose
+            # We'll insert a node corresponding to our current position.
+            # Difficulty is determining it's adjacency to other nodes, based on geometry
+            robot_name = rospy.get_namespace().split("/")[-2]
```

Comment:
This should be done in the constructor. Also I am sure this isn't robust, as if run this in other namespaces, `-2` is not the first anymore.

Reply:
I have no alternative AFAIK

Reply:
It is a list, so why not take the first item?